tiledb/sm/cpp_api/filter.h

```diff
@@ -335,67 +335,84 @@ class Filter {
   template <typename T>
   void option_value_typecheck(tiledb_filter_option_t option) {
     std::string type_name = tiledb::impl::type_to_tiledb<T>().name;
+    const char* option_name;
+    tiledb_filter_option_to_str(option, &option_name);
     switch (option) {
       case TILEDB_COMPRESSION_LEVEL:
-        if constexpr (!std::is_same_v<int32_t, T>)
+        if constexpr (!std::is_same_v<int32_t, T>) {
           throw std::invalid_argument(
-              "Cannot set option with type '" + type_name +
-              "'; Option value must be int32_t.");
+              "Cannot set filter option '" + std::string(option_name) +
+              "' with type '" + type_name + "'; Option value must be '" +
+              tiledb::impl::type_to_tiledb<int32_t>().name + "'.");
+        }
```

Comment:
There's an awful lot of copypasta in these error messages. All but the last one, which needs a custom message, could go into a purpose-built exception class. This is longer than it needs to be, since the private elements should ideally be defined already.
```
  template <typename T, typename U>
  class invalid_filter_option : public std::invalid_argument {
    static std::string option_name(tiledb_filter_option_t option) {
      const char* option_name;
      tiledb_filter_option_to_str(option, &option_name);
      return {option_name};
    }
    static constexpr char* argument_type =
        tiledb::impl::type_to_tiledb<T>().name;
    static constexpr char* expected_type =
        tiledb::impl::type_to_tiledb<U>().name;

   public:
    invalid_filter_option(tiledb_filter_option_t option)
        : std::invalid_argument(
              "Cannot set filter option '" + option_name(option) +
              "' with type '" + argument_type + "'; Option value must be '" +
              expected_type + "'.") {
    }
  };
```

It'd also be worth defining
```
namespace tiledb::impl{
template <typename T>
constexpr const char* type_to_tiledb_name = type_to_tiledb<T>().name;
}
```
in `type.h`, as well as an analogous one for options. No need to include them in the scope of this PR unless feeling inspired.

Reply:
@eric-hughes-tiledb If you feel strongly I can add the error class now. Just let me know so I can wrap this PR up.

Reply:
>If you feel strongly

Yes. We are strongly discouraging copypasta in new code where it can be avoided, and it can be avoided here.

>the most error prone part [...] (matching correct types)

That's the irreducible part of the specification. There's an expected type as the template argument to `option_value_typecheck`, and there's an actual type present in the option. Unfortunately C++ does not provide a named argument syntax, which would make things clearer. It might help to rename various `T` and `U` formal arguments to `Expected` and `Actual` as appropriate.

>I think your suggested error class belongs with some such internal infrastructure for the FilterOption enum

That would not work well here. This is in the C++ API and we cannot access internal infrastructure.

It's also not a great idea to simply forward an error message up through the C API. This is also the C++ representation of the type system, not the internal one; the names are not identical, if nothing else.

Reply:
> That's the irreducible part of the specification.

I have some ideas of how to do it, but it's way beyond the scope of this PR.

> That would not work well here. This is in the C++ API and we cannot access internal infrastructure.

Internal as in part of an `impl` namespace in the C++ API.